### PR TITLE
Add IPC security section

### PR DESCRIPTION
**Pull request by JamesNK**
@blowdart @GrabYourPitchforks Security guidance for named pipes IPC (and IPC in general)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/grpc/interprocess.md](https://github.com/dotnet/AspNetCore.Docs/blob/ac40eebc2ca99d8eb173151c88ea6791c4a05dd3/aspnetcore/grpc/interprocess.md) | [Inter-process communication with gRPC](https://review.learn.microsoft.com/en-us/aspnet/core/grpc/interprocess?branch=pr-en-us-29973) |

### CODE REVIEW

File: aspnetcore/grpc/interprocess.md [hunk 1]
```diff
@@ -69,6 +69,71 @@ builder.WebHost.ConfigureKestrel(serverOptions =>
...
+### Secure IPC server app against unexpected callers
+
+The IPC server app hosts RPC services for other apps to call. Incoming callers should be authenticated to prevent untrusted clients from making RPC calls to the server.
+
+Transport security is one option for securing a server. IPC transports such as Unix domain sockets and named pipes support limiting access based on operating system permissions:
+
+* Named pipes supports securing a pipe with the [Windows access control model](https://learn.microsoft.com/windows/win32/ipc/named-pipe-security-and-access-rights). Access rights can be configured in .NET when a server is started using the <xref:System.IO.Pipes.PipeSecurity> class.
+* Unix domain sockets support securing a socket with file permissions.
+
+Another option for securing a IPC server is to use authentication and authorization built into ASP.NET Core. For example, the server could be configured to require [certificate authentication](https://learn.microsoft.com/aspnet/core/security/authentication/certauth). RPC calls made by client apps without the require certificate fail with an unauthorized response.
```
Comment (gfoidl) [inline aspnetcore/grpc/interprocess.md:85]:
```suggestion
Another option for securing a IPC server is to use authentication and authorization built into ASP.NET Core. For example, the server could be configured to require [certificate authentication](https://learn.microsoft.com/aspnet/core/security/authentication/certauth). RPC calls made by client apps without the required certificate fail with an unauthorized response.
```

File: aspnetcore/grpc/interprocess.md [hunk 2]
```diff
@@ -69,6 +69,71 @@ builder.WebHost.ConfigureKestrel(serverOptions =>
...
+
+The IPC server app hosts RPC services for other apps to call. Incoming callers should be authenticated to prevent untrusted clients from making RPC calls to the server.
+
+Transport security is one option for securing a server. IPC transports such as Unix domain sockets and named pipes support limiting access based on operating system permissions:
+
+* Named pipes supports securing a pipe with the [Windows access control model](https://learn.microsoft.com/windows/win32/ipc/named-pipe-security-and-access-rights). Access rights can be configured in .NET when a server is started using the <xref:System.IO.Pipes.PipeSecurity> class.
+* Unix domain sockets support securing a socket with file permissions.
+
+Another option for securing a IPC server is to use authentication and authorization built into ASP.NET Core. For example, the server could be configured to require [certificate authentication](https://learn.microsoft.com/aspnet/core/security/authentication/certauth). RPC calls made by client apps without the required certificate fail with an unauthorized response.
+
```
Comment (blowdart) [inline aspnetcore/grpc/interprocess.md:86]: Commit to making sure this advice is on the docs page for enabling IPC

File: aspnetcore/grpc/interprocess.md [hunk 3]
```diff
@@ -69,6 +69,71 @@ builder.WebHost.ConfigureKestrel(serverOptions =>
...
+            if (sslPolicyErrors != SslPolicyErrors.None)
+            {
+                return false;
+            }
+
+            // Validate server cert thumbprint matches the expected thumbprint.
+        }
+    }
+};
+```
```
Comment (blowdart) [inline aspnetcore/grpc/interprocess.md:121]: Needs public docs before release

File: aspnetcore/grpc/interprocess.md [hunk 4]
```diff
@@ -69,6 +69,71 @@ builder.WebHost.ConfigureKestrel(serverOptions =>
...
+Client's can protect against this attack by not allowing impersonation when connecting to a server. Unless required by a server, a <xref:System.Security.Principal.TokenImpersonationLevel> value of `None` or `Anonymous` should be used when creating a client connection:
+
+```cs
+using var pipeClient = new NamedPipeClientStream(
+    serverName: ".", pipeName: "testpipe", PipeDirection.In, PipeOptions.None, TokenImpersonationLevel.None);
+await pipeClient.ConnectAsync();
+```
+
+`TokenImpersonationLevel.None` is the default value in `NamedPipeClientStream` constructors that don't have an `impersonationLevel` parameter.
+
```
Comment (blowdart) [inline aspnetcore/grpc/interprocess.md:136]: And you know already, describe this publically

**Review by blowdart — APPROVED** [review]

File: aspnetcore/grpc/interprocess.md [hunk 5]
```diff
@@ -69,6 +69,71 @@ builder.WebHost.ConfigureKestrel(serverOptions =>
...
+IPC apps send and receive RPC calls. External communication is a potential attack vector for IPC apps and must be properly secured.
+
+### Secure IPC server app against unexpected callers
+
+The IPC server app hosts RPC services for other apps to call. Incoming callers should be authenticated to prevent untrusted clients from making RPC calls to the server.
+
+Transport security is one option for securing a server. IPC transports such as Unix domain sockets and named pipes support limiting access based on operating system permissions:
+
+* Named pipes supports securing a pipe with the [Windows access control model](https://learn.microsoft.com/windows/win32/ipc/named-pipe-security-and-access-rights). Access rights can be configured in .NET when a server is started using the <xref:System.IO.Pipes.PipeSecurity> class.
```
Comment (Rick-Anderson) [inline aspnetcore/grpc/interprocess.md:82]:
```suggestion
* Named pipes supports securing a pipe with the [Windows access control model](/windows/win32/ipc/named-pipe-security-and-access-rights). Access rights can be configured in .NET when a server is started using the <xref:System.IO.Pipes.PipeSecurity> class.
```

File: aspnetcore/grpc/interprocess.md [hunk 6]
```diff
@@ -69,6 +69,71 @@ builder.WebHost.ConfigureKestrel(serverOptions =>
...
+* Unix domain sockets support securing a socket with file permissions.
+
+Another option for securing a IPC server is to use authentication and authorization built into ASP.NET Core. For example, the server could be configured to require [certificate authentication](https://learn.microsoft.com/aspnet/core/security/authentication/certauth). RPC calls made by client apps without the required certificate fail with an unauthorized response.
```
Comment (Rick-Anderson) [inline aspnetcore/grpc/interprocess.md:85]:
```suggestion
Another option for securing a IPC server is to use authentication and authorization built into ASP.NET Core. For example, the server could be configured to require [certificate authentication](xref:security/authentication/certauth). RPC calls made by client apps without the required certificate fail with an unauthorized response.
```

File: aspnetcore/grpc/interprocess.md [hunk 7]
```diff
@@ -69,6 +69,71 @@ builder.WebHost.ConfigureKestrel(serverOptions =>
...
+internal static bool CheckPipeConnectionOwnership(
+    NamedPipeClientStream pipeStream, SecurityIdentifier expectedOwner)
+{
+    var remotePipeSecurity = pipeStream.GetAccessControl();
+    var remoteOwner = remotePipeSecurity.GetOwner(typeof(SecurityIdentifier));
+    return expectedOwner.Equals(remoteOwner);
+}
+```
+
+Another option for validating the server is to [secure its endpoints with HTTPS](https://learn.microsoft.com/aspnet/core/fundamentals/servers/kestrel/endpoints#configure-https) inside ASP.NET Core. The client can configure `SocketsHttpHandler` to validate the server is using the expected certificate when the connection is established.
```
Comment (Rick-Anderson) [inline aspnetcore/grpc/interprocess.md:103]:
```suggestion
Another option for validating the server is to [secure its endpoints with HTTPS](/aspnet/core/fundamentals/servers/kestrel/endpoints#configure-https) inside ASP.NET Core. The client can configure `SocketsHttpHandler` to validate the server is using the expected certificate when the connection is established.
```

File: aspnetcore/grpc/interprocess.md [hunk 8]
```diff
@@ -69,6 +69,71 @@ builder.WebHost.ConfigureKestrel(serverOptions =>
...
+
+            // Validate server cert thumbprint matches the expected thumbprint.
+        }
+    }
+};
+```
+
+### Protect against named pipe privilege escalation
+
+Named pipes supports a feature called [impersonation](https://learn.microsoft.com/windows/win32/ipc/impersonating-a-named-pipe-client). Using impersonation, the named pipes server can execute code with the privileges of the client user. This is a powerful feature but can allow a low-privilege server to impersonate a high-privilege caller and then run malicious code.
```
Comment (Rick-Anderson) [inline aspnetcore/grpc/interprocess.md:125]:
```suggestion
Named pipes supports a feature called [impersonation](/windows/win32/ipc/impersonating-a-named-pipe-client). Using impersonation, the named pipes server can execute code with the privileges of the client user. This is a powerful feature but can allow a low-privilege server to impersonate a high-privilege caller and then run malicious code.
```

**Review by Rick-Anderson — CHANGES_REQUESTED** [review]

**Review by Rick-Anderson — APPROVED** [review]

**Pull request merged**